File: dotNet/Tests/ContextBuilderTests.cs
using System;
using NUnit.Framework;

namespace Com.Company.Context {
    public class TestConfig {
        [Bean]
        public string GetInstance() {
            return "Instance1";
        }

        [Bean("bean3")]
        public string GetInstance3([BeanParameter("bean1")]string input) {
            return "Instance1";
        }

        [Bean("bean1")]
        public string GetInstance1() {
            return "Instance1";
        }

        [Bean("bean2")]
        public string GetInstance2([BeanParameter("bean1")]string input) {
            return "Instance1";
        }
    }

    public class EmptyConfig {
        public string GetInstance() {
            return "Instance";
        }
    }

    public class VoidBeanType {
        [Bean("bean2")]
        public void GetVoid() {}
    }

    public class IncompleteConfig {
        [Bean("bean2")]
        public string GetInstance2([BeanParameter("bean1")]string input) {
            return "Instance1";
        }
    }

    public class NullBeanId {
        [Bean(null)]
        public string GetInstance2([BeanParameter("bean1")]string input) {
            return "Instance1";
        }
    }

    public class NullBeanParamId {
        [Bean]
        public string GetInstance2([BeanParameter(null)]string input) {
            return "Instance1";
        }
    }

    public class MultipleTypes {
        [Bean]
        public object GetInstance1() { return new object(); }

        [Bean]
        public object GetInstance2() { return new object(); }
    }

    public class MultipleNamesDiffTypes {
        [Bean("bean1")]
        public object GetInstance1() { return new object(); }

        [Bean("Bena2")]
        public object GetInstance2() { return new object(); }
    }

    public class ContextBuilderTests {
        [Test]
        public void MultipleNamesDiffTypes_Test() {
            try {
                var builder = new ContextBuilder();
                var context = builder.Build(new MultipleNamesDiffTypes());
                Assert.NotNull(context);
            } catch (IllegalBeanException) {
            }
        }

        [Test]
        public void MultipleTypes_Test() {
            try {
                var builder = new ContextBuilder();
                builder.Build(new MultipleTypes());
                Assert.Fail();
            } catch (IllegalConfigException) {
            }
        }

        [Test]
        public void CreateSimple_Test() {
            var builder = new ContextBuilder();
            var context = builder.Build(new TestConfig());
            Assert.NotNull(context);
        }

        [Test]
        public void VoidBeanType_Test() {
            try {
                var builder = new ContextBuilder();
                builder.Build(new VoidBeanType());
                Assert.Fail();
            } catch (IllegalBeanException) {
            }
        }

        [Test]
        public void NullConfig_Test() {
            try {
                var builder = new ContextBuilder();
                builder.Build(null);
                Assert.Fail();
            } catch (IllegalConfigException) {
            }
        }

        [Test]
        public void EmptyConfig_Test() {
            try {
                var builder = new ContextBuilder();
                builder.Build(new EmptyConfig());
                Assert.Fail();
            } catch (IllegalConfigException) {
            }
        }

        [Test]
        public void AbsentParamBean_Test() {
            try {
                var builder = new ContextBuilder();
                builder.Build(new IncompleteConfig());
                Assert.Fail();
            } catch (IllegalBeanException) {
            }
        }

        [Test]
        public void NullBeanId_Test() {
            try {
                var builder = new ContextBuilder();
                builder.Build(new NullBeanId());
                Assert.Fail();
            } catch (NullReferenceException) {
            }
        }

        [Test]
        public void NullBeanParamId_Test() {
            try {
                var builder = new ContextBuilder();
                builder.Build(new NullBeanParamId());
                Assert.Fail();
            } catch (NullReferenceException) {
            }
        }
    }
}

File: java/tests/com/company/context/ContextBuilderTests.java
package com.company.context;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.fail;

import org.junit.jupiter.api.Test;

import java.lang.reflect.InvocationTargetException;

class ContextBuilderTests {
    @Test
    void createSimple_Test() throws InvocationTargetException, IllegalAccessException, IllegalConfigException, IllegalBeanException {
        Object config = new Object() {
            @Bean
            public String getObject() { return "Instance"; }

            @Bean(id = "bean3")
            public String getObject1(@BeanParameter(id = "bean1") String input) {
                return "Instance";
            }

            @Bean(id = "bean1")
            public String getObject2() { return "Instance"; }

            @Bean(id = "bean2")
            public String getObject3(@BeanParameter(id = "bean1") String input) {
                return "Instance";
            }
        };

        IContextBuilder builder = new ContextBuilder();

        IContext context = builder.build(config);
        assertNotNull(context, "Context is null");
    }

    @Test
    void voidBean_Test() throws IllegalAccessException, IllegalBeanException, InvocationTargetException {
        Object config = new Object() {
            @Bean
            public String getObject1() { return "Instance 1"; }

            @Bean
            public void getObject2() { }
        };

        IContextBuilder builder = new ContextBuilder();

        try {
            builder.build(config);
            fail();
        } catch (IllegalConfigException ex) {
        }
    }

    @Test
    void multipleType_Test() throws IllegalAccessException, IllegalBeanException, InvocationTargetException {
        Object config = new Object() {
            @Bean
            public String getObject1() { return "Instance 1"; }

            @Bean
            public String getObject2() { return "Instance 2"; }
        };

        IContextBuilder builder = new ContextBuilder();

        try {
            builder.build(config);
            fail("Multiple unnamed instances with same type are not allowed");
        } catch (IllegalConfigException ex) {
        }
    }

    @Test
    void multipleNamedDiffType_Test() throws IllegalAccessException, IllegalBeanException, InvocationTargetException, IllegalConfigException {
        Object config = new Object() {
            @Bean(id = "object")
            public String getObject1() { return "Instance 1"; }

            @Bean(id = "object")
            public Object getObject2() { return new Object(); }
        };

        IContextBuilder builder = new ContextBuilder();

        builder.build(config);
    }

    @Test
    void nullConfig_Test() throws IllegalAccessException, IllegalBeanException, InvocationTargetException {
        IContextBuilder builder = new ContextBuilder();

        try {
            builder.build(null);
            fail("Null config are not allowed");
        } catch (IllegalConfigException ex) {
        }
    }

    @Test
    void emptyConfig_Test() throws IllegalAccessException, IllegalBeanException, InvocationTargetException {
        IContextBuilder builder = new ContextBuilder();

        try {
            Object config = new Object();
            builder.build(config);
            fail("Empty config are not allowed");
        } catch (IllegalConfigException ex) {
        }
    }

    @Test
    void absentParamBean_Test() {
        Object config = new Object() {
            @Bean(id = "inst")
            public String getObject1(@BeanParameter(id = "val3")int obj) { return "Instance " + obj; }

            @Bean(id = "val1")
            public int getVal1() { return 10; }

            @Bean(id = "val2")
            public int getVal2() { return 20; }
        };

        IContextBuilder builder = new ContextBuilder();

        try {
            builder.build(config);
            fail("Absent bean param is not allowed");
        } catch (Exception ex) {
            assertEquals(ex.getClass(), IllegalBeanException.class);
        }
    }
}